### ISSUE TITLE
Automatically make a 'parent' resource 'dependent' on its child resources completing.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [pulumi/pulumi#2295]
Conclusions from discussion between @lukehoban @swgillespie @pgavlin and @CyrusNajmabadi:

We discussed a general problem that exists today that is being exacerbated by teh work currently being done in `awsinfra`.  Specifically, we commonly want to make component resources that logically represent a set of physical custom resources.  We want other components to be able to 'dependsOn' these components with the semantics that that means the 'dependsOn' should only be 'done' when the *logibal* resource is 'done' (which means all its physical dependencies are 'done').

Currently, for a component resource to be 'done' it's 'urn' must be complete.  As such, in many cases, we make this work by manually having the component replace it's urn with something akin to:

```ts
super(...);

const child1 = new aws...SomeResource(..., { parent: this });
const child2 = new aws...OtherResource(..., { parent: this });

const oldUrn = this.urn;
const allDeps = new Set(...this.urn.resources());
allDeps.add(child1);
allDeps.add(child2);
this.urn = new Output(allDeps, this.urn.promise();
```

This is pretty inconvenient as this must be done manually and kept up to date as things change.

We propose making the above automatic.  Specifically, when a resource is created and passed another resource as it's 'parent', then that resource will augment/decorate the 'urn' for the parent resource in this fashion.  By doing this, once the parent resource is complete, it will inherently be dependent on all the children it created in its constructor.

Note though that we will also happen if this resource is used as the parent of some other resource later on (which is common in awsinfra when you call a method on a resource asking it to make a new resource that can logically be felt to be a child).

--

On top of this, we will also add a facility for a Resource to add something htat should be considered a dependent even for the case where it woudl not be the 'parent' of that resource.  this also happens in awsinfra and is the not-uncommon case of making an 'attachment' resource that then wants the resource it is attached to to no longer be 'done' until it is done as well.  

Note: this would effectively be the mechanism by which a child would register itself as a dependency of hte parent.  Not 100% sure what the shape will be, but it will likely be:

```ts
class Resource {
//...
    public attachDependency(promise: Promise);
    public attachDependency(resource: Resource);
```

The first method is valuable as a general approach of allowing a Resource to say it is only done when some async work is done (like a health check).  The latter is for explicitly stating it's not done until some other Resource is done.  Note: the latter may actually be unnecessary.  it's likely that the latter is nothing more than a shortcut for `this.attachDependency(res.urn.promise())`.

**Comment by lukehoban** [comment]
@CyrusNajmabadi Worth taking a look at #1743 and #994 to understand to what extent the proposal here will address all or some of the issues in those two.

**Comment by lukehoban** [comment]
We explored a possible solution earlier in this milestone - of modifying the `urn` promise (and possible also all other output property promises) whenever a child was added to wait on the child as well before resolving.  This ran into some subtle problems with timing of when parts of the system read this state, and generally with the explicit state-fullness it exposes.  We were also concerned about possibilities for deadlocks.

Given these problems, we'll need to revisit a design approach here at a potentially deeper level.
